File: Construct Binary Tree from Preorder and Inorder Traversal.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        // Start typing your Java solution below
        // DO NOT write main() function
        HashMap<Integer, Integer> indexMap = new HashMap<Integer, Integer>();
        for(int i = 0; i < inorder.length; i++)
        {
            indexMap.put(inorder[i], i);
        }
        
        return buildTree(preorder, preorder.length, 0, 0, indexMap);
    }
    
    public TreeNode buildTree(int[] preorder, int size, int start, int offset, HashMap<Integer, Integer> indexMap)
    {
        if(size < 1)
        {
            return null;
        }
        int val = preorder[start];
        TreeNode root = new TreeNode(val);
        int index = indexMap.get(val);
        int leftSize = index - offset;
        int rightSize = size - leftSize - 1;
        root.left = buildTree(preorder, leftSize, start + 1, offset, indexMap);
        root.right = buildTree(preorder, rightSize, start + leftSize + 1, offset + leftSize + 1,  indexMap);
        return root;
    }
    
    
}

File: Construct Binary Tree from Preorder and Inorder Traversal.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    
    unordered_map<int, int> indexMap;
    
    TreeNode *buildTree(vector<int> &preorder, vector<int> &inorder) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        
        //this->preorder = preorder;
        //this->inorder = inorder;
        
        int N = preorder.size();
        indexMap.clear();
        for(int i = 0; i < N; i++)
        {
            indexMap[inorder[i]] = i;
        }
        
        return buildTree(preorder, 0, N, 0);
    }
    
    TreeNode *buildTree(vector<int> &preorder, int preorderStart, int N, int offset)
    {
        if(N<=0)
            return NULL;
        
        int val = preorder[preorderStart];
        int inorderIndex = indexMap[val];
        TreeNode *root = new TreeNode(val);
        int i = inorderIndex - offset;
        
        root->left = buildTree(preorder, preorderStart+1, i, offset);
        root->right = buildTree(preorder, preorderStart+i+1, N-i-1, offset + i + 1);
        
        return root;
    }
    
};